File: my-code/0337-house-robber-iii/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    pair<int,int> dfs(TreeNode* root) //first->withroot max,second->withoutroot max
    {
        if(root==NULL)
        {
            return {0,0};
        }
        pair<int,int> leftpair= dfs(root->left);
        pair<int,int> rightpair= dfs(root->right);
        int left=root->val+leftpair.second+rightpair.second;
        int right=max(leftpair.first,leftpair.second)+max(rightpair.first,rightpair.second);
        return {left,right};
    }

    int rob(TreeNode* root) {
        pair<int,int> ans=dfs(root);
        return max(ans.first,ans.second);
    }
};

File: my-code/0337-house-robber-iii/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class pair
{
  int first;int second;
   pair(int left,int right)
  {
    this.first=left;
    this.second=right;
  }
}

class Solution {
    pair dfs(TreeNode root) //first->withroot max,second->withoutroot max
    {
        if(root==null)
        {
            return new pair(0,0);
        }
        pair leftpair= dfs(root.left);
        pair rightpair= dfs(root.right);
        int left=root.val+leftpair.second+rightpair.second;
        int right=Math.max(leftpair.first,leftpair.second)+Math.max(rightpair.first,rightpair.second);
        return new pair(left,right);
    }
    public int rob(TreeNode root) {
        pair ans=dfs(root);
        return Math.max(ans.first,ans.second);
    }
}
